File: Permutations of a given string - GFG/permutations-of-a-given-string.cpp
//{ Driver Code Starts
#include<bits/stdc++.h>
using namespace std;

// } Driver Code Ends
class Solution
{
	public:
	    void permute(string& a, int l, int r, vector<string> &res){
            if (l == r)
                res.emplace_back(a);
            else {
                for (int i = l; i <= r; i++) {
                    swap(a[l], a[i]);
                    permute(a, l + 1, r,res);
                    swap(a[l], a[i]);
                }
            }
        }
		vector<string>find_permutation(string S)
		{
		    // Code here there
		    vector<string>res;
		    int n = S.size();
		    permute(S, 0, n-1, res);
		    map<string, int>mp;
		    for(int i = 0; i<res.size();i++){
		        mp[res[i]]++;
		    }
		    res.clear();
		    for(auto it: mp){
		        res.emplace_back(it.first);
		    }
		    return res;
		}
};



//{ Driver Code Starts.
int main(){
    int t;
    cin >> t;
    while(t--)
    {
	    string S;
	    cin >> S;
	    Solution ob;
	    vector<string> ans = ob.find_permutation(S);
	    for(auto i: ans)
	    {
	    	cout<<i<<" ";
	    }
	    cout<<"\n";
    }
	return 0;
}

// } Driver Code Ends

File: Permutations of a given string - GFG/permutations-of-a-given-string.py
#User function Template for python3
import math as m
class Solution:
    def find_permutation(self, S):
        # Code here
        def per(s):
            ind = -1
            for i in range(len(s)-2, -1, -1):
                if s[i] < s[i+1]:
                    ind = i
                    break
            if ind == -1:
                s = s[::-1]
                return s
            for i in range(len(s)-1, ind, -1):
                if s[i] > s[ind]:
                    s = s[:ind] + s[i] + s[ind+1:i] + s[ind] + s[i+1:]
                    break
            s = s[:ind+1] + s[ind+1:][::-1]
            return s
        
        n = m.factorial(len(S))
        res =[S]
        while(n):
            p = S
            S = per(p)
            if S not in res:
                res.append(S)
            n -= 1
        return sorted(res)
        
#{ 
 # Driver Code Starts
#Initial Template for Python 3


if __name__ == '__main__':
	t=int(input())
	for i in range(t):
		S=input()
		ob = Solution()
		ans = ob.find_permutation(S)
		for i in ans:
			print(i,end=" ")
		print()
# } Driver Code Ends